[20725] Add new watcher handling

floriank opened the pull request:
This PR will add improved handling for the watchers tab in the split screen view.

It should cover the needs of https://community.openproject.org/work_packages/20725.
## Things to discuss/todo:
- **TESTS**
  - [x] basic tests for services and controller
- [x] Styling is not done at all (cc @HDinger)
- [x] String are not translated at the moment

As always, feedback welcome

Commit:
initial splitoff of watchers into separate service

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
prepare the directives and load both available and watching users

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add a lookup directive

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
provide a method of adding watchers

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make watchers removable

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
remove dead code and provide helpers for the array sets

Signed-off-by: Florian Kraft <f.kraft@finn.de>

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://floriank-feature-20725-le-awesome-watchers-implementation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-feature-20725-le-awesome-watchers-implementation.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

floriank commented on frontend/app/templates/work_packages/watchers/watcher.html:
```diff
@@ -0,0 +1,7 @@
+<li data-ng-class="{'loading':watcher.loading,'deleting':deleting}">
+  <a ng-href="{{watcher._links.self.href}}" title="{{watcher.name}}">
```
**NOTE** to myself: this is the API response and **not** the users profile page

Commit:
handle injections to watchers service properly and add tests

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
do not only tests the WatchersService

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
remove old controller tests for watchers tab

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add some basic tests

These tests are mostly concerned with the data layer and expected and
actual responses from the APIv3.

Also, a controller skeleton test is included

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
insert translations

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
@ulferts @basiszwo I could add tests for integration and directives, however that will take time - should I?

floriank commented:
@HDinger I'd like to delegate styling to you - if you have questions, we can talk.

floriank commented:
@opf/developers In general, I'd like feedback on this, I chose a more pragmatical approach in implementation as time is limited, so I could use all of your input.

Commit:
use correct path for linking users profile

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
remove focus statement

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
Notes from @basiszwo (via email, yesterday evening):
- watcher selection should be a `ui.select`, like the assignee select
  - on click the list of available users should be shown (default behaviour)
- the mini menu (like with the other in place edit fields should be shown to confirm the selection, but without the "send email" button
- the user is added above in the list (**NOTE to self:** _Appended_ to the list above, not _prepended_)

Commit:
use ui.select for watcher selection

This will use the ui.select based implementation already forund in the
work package form, however, this introduces a more brittle copy-pasted
approach to simulate the appearance.

an actual view component is not included

Commit:
use a several parts from inplace edit to change add watcher workflow

Watchers can now be added like assignees. The component now has its own
implementation of the trigger component, based on edit and display pane
of the work package form.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

HDinger committed:
styled the watchers tab

HDinger committed:
changed the styling of attachments in details pane similar to the watchers

ulferts commented:
@HDinger I have the impression that the height of the watchers is less than the height of the inplace edit fields. Can you make them have the same height?

Watchers
![image](https://cloud.githubusercontent.com/assets/617519/9406884/a937f212-4804-11e5-9d59-f6d72e0a770c.png)
vs Inplace edit fields:
![image](https://cloud.githubusercontent.com/assets/617519/9406896/bfda0136-4804-11e5-8ca3-340484c205c0.png)

ulferts commented on app/assets/stylesheets/content/work_package_details/_watchers_tab.sass:
```diff
@@ -38,6 +36,9 @@
   input[type='text']
     width: 100%
```
Do we still need all those styles @HDinger. Especially the select2/ui-select once I assume to be unnecessary by now.

ulferts commented:
Just a note to self or whoever will take over the branch. It would probably be good to double check (ask the user) before deleting an attachment. I don't think that it is necessary for watchers as the user is not actually deleted, but an attachment is gone once the user clicks the icon.

ulferts commented on frontend/webpack.config.js:
```diff
@@ -124,7 +124,8 @@ module.exports = {
         'openproject-ui_components/app/assets/javascripts/angular/ui-components-app',
-      'ngFileUpload': 'ng-file-upload/ng-file-upload'
+      'ngFileUpload': 'ng-file-upload/ng-file-upload',
+      'angucomplete-alt': 'angucomplete-alt/angucomplete-alt'
```
This is a relic introduced in a9538ca3 when `angucomplete` was still used in lookup.html.

floriank replied:
This should be easily removable as I already removed the dependencies in `openproject-core-app.js`

HDinger committed:
fixed line height of the new fields and removed the watchers tab from the styleguide

HDinger commented:
During the styling some functionality got lost in the attachment field. At the moment you cannot load the attachment by directly clicking on it.

ulferts commented:
Superseeded by #3434
